pkg/providers/amifamily/al2023.go

```diff
@@ -15,38 +15,72 @@ limitations under the License.
 package amifamily
 
 import (
+	"context"
 	"fmt"
+	"regexp"
+	"strings"
 
 	"github.com/samber/lo"
 	v1 "k8s.io/api/core/v1"
 	corev1beta1 "sigs.k8s.io/karpenter/pkg/apis/v1beta1"
 	"sigs.k8s.io/karpenter/pkg/cloudprovider"
 	"sigs.k8s.io/karpenter/pkg/scheduling"
 
+	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
 	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily/bootstrap"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/ssm"
 )
 
 type AL2023 struct {
 	DefaultFamily
 	*Options
 }
 
-func (a AL2023) DefaultAMIs(version string) []DefaultAMIOutput {
-	return []DefaultAMIOutput{
-		{
-			Query: fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2023/x86_64/standard/recommended/image_id", version),
-			Requirements: scheduling.NewRequirements(
-				scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, corev1beta1.ArchitectureAmd64),
-			),
-		},
-		{
-			Query: fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2023/arm64/standard/recommended/image_id", version),
-			Requirements: scheduling.NewRequirements(
-				scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, corev1beta1.ArchitectureArm64),
-			),
-		},
+func (a AL2023) AMIQuery(ctx context.Context, ssmProvider ssm.Provider, k8sVersion string, amiVersion string) (AMIQuery, error) {
+	query := AMIQuery{
+		Filters: []*ec2.Filter{&ec2.Filter{
+			Name: lo.ToPtr("image-id"),
+		}},
+		KnownRequirements: make(map[string][]scheduling.Requirements),
+	}
+	rootPath := fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2023", k8sVersion)
```

Comment:
This approach minimizes the calls to `GetParametersByPath` by getting the results for all variants (including future GPU variants) in a single call. Since AL2 uses different root paths we need to do multiple calls, but we can get away with a single call here. Personally I think this is the approach we should take when possible.

Reply:
I think you're right that this is the best we can do now.